File: Java/DesignPatterns/Subject.java
// Part of the Observer demo.

public class Subject
{
    // The root of the subscribed objects in the linked-list structure:
    Observer root;

    public Subject()
    {
        root = null;
    }

    // Add an observer to the list:
    // You can also feel free to use a built-in structure like a List<T> instead!
    public void AddObserver(Observer observer)
    {
        // Insert head-first into the list.
        observer.next = root;
        root = observer;
    }

    public void Notify(String message)
    {
        // Iterate the observers, notifying them all:
        Observer tmp = this.root;
        while(tmp != null)
        {
            tmp.Update(message);
            tmp = tmp.next;
        }
    }

}

File: CSharp/DesignPatterns/Subject.cs
// Part of the observer demo
using System.Collections.Generic;

public class StockTicker
{
    // For brevity, we'll be using a built-in List structure:
    List<TickerObserver> observers;

    public StockTicker()
    {
        observers = new List<TickerObserver>();
    }

    public void AddObserver(TickerObserver observer)
    {
        observers.Add(observer);
    }

    public void RemoveObserver(TickerObserver observer)
    {
        observers.Remove(observer);
    }

    // In our design, we're using one observer for all stock symbols.
    // This design can be improved on further by designating one subject per ticker symbol!
    public void Notify(string tickerSymbol, float price)
    {
        // Notify all the observers of the price update.
        foreach(TickerObserver observer in observers)
        {
            observer.Update(tickerSymbol, price);
        }
    }
}